src/interface.jl

```diff
@@ -0,0 +1,252 @@
+    ADBACKEND[] = backend_sym
+end
+
+ADBackend() = ADBackend(ADBACKEND[])
+ADBackend(T::Symbol) = ADBackend(Val(T))
+function ADBackend(::Val{T}) where {T}
+    if T === :forward_diff
+        return ForwardDiffAD
+    else
+        return TrackerAD
+    end
+end
+
+######################
+# Bijector interface #
+######################
+
+abstract type Bijector end
+abstract type ADBijector{AD} <: Bijector end
+
+struct Inversed{B <: Bijector} <: Bijector
+    orig::B
+end
+
+Broadcast.broadcastable(b::Bijector) = Ref(b)
+
+"Computes the log(abs(det(J(x)))) where J is the jacobian of the transform."
+logabsdetjac(b::T1, y::T2) where {T<:Bijector,T1<:Inversed{T},T2} = 
+    error("`logabsdetjac(b::$T1, y::$T2)` is not implemented.")
+
+"Transforms the input using the bijector."
+transform(b::T1, y::T2) where {T<:Bijector,T1<:Inversed{T},T2} =
+    error("`transform(b::$T1, y::$T2)` is not implemented.")
+
+"Computes both `transform` and `logabsdetjac` in one forward pass."
+forward(b::T1, y::T2) where {T<:Bijector,T1<:Inversed{T},T2} = 
+    error("`forward(b::$T1, y::$T2)` is not implemented.")
+
+
+transform(b::Bijector) = x -> transform(b, x)
+(ib::Inversed{<: Bijector})(y) = transform(ib, y)
+
+# default `forward` implementations; should in general implement efficient way
+# of computing both `transform` and `logabsdetjac` together.
+forward(b::Bijector, x) = (rv=transform(b, x), logabsdetjac=logabsdetjac(b, x))
+forward(ib::Inversed{<: Bijector}, y) = (rv=transform(ib, y), logabsdetjac=logabsdetjac(ib, y))
+
+# defaults implementation for inverses
+logabsdetjac(ib::Inversed{<: Bijector}, y) = - logabsdetjac(ib.orig, transform(ib, y))
+
+inv(b::Bijector) = Inversed(b)
+inv(ib::Inversed{<:Bijector}) = ib.orig
+
+# AD implementations
+function jacobian(b::ADBijector{<: ForwardDiffAD}, y::Real)
+    ForwardDiff.derivative(z -> transform(b, z), y)
+end
+function jacobian(b::Inversed{<: ADBijector{<: ForwardDiffAD}}, y::Real)
+    ForwardDiff.derivative(z -> transform(b, z), y)
+end
+function jacobian(b::Inversed{<: ADBijector{<: ForwardDiffAD}}, y::AbstractVector{<: Real})
+    ForwardDiff.jacobian(z -> transform(b, z), y)
+end
+
+function jacobian(b::ADBijector{<: TrackerAD}, y::Real)
+    Tracker.gradient(z -> transform(b, z), y)[1]
+end
+function jacobian(b::Inversed{<: ADBijector{<: TrackerAD}}, y::Real)
+    Tracker.gradient(z -> transform(b, z), y)[1]
+end
+function jacobian(b::Inversed{<: ADBijector{<: TrackerAD}}, y::AbstractVector{<: Real})
+    Tracker.jacobian(z -> transform(b, z), y)
+end
+
+# TODO: allow batch-computation, especially for univariate case?
+"Computes the absolute determinant of the Jacobian of the inverse-transformation."
+function logabsdetjac(b::ADBijector, x::Real)
+    log(abs(jacobian(b, x)))
+end
+function logabsdetjac(b::ADBijector, x::AbstractVector{<:Real})
+    fact = lu(jacobian(b, x), check=false)
+    issuccess(fact) ? log(abs(det(fact))) : -Inf # TODO: or smallest possible float?
```

Comment:
1. Return please :)
2. Maybe we want to error if this fails, rather than throw a big large number?

Reply:
It was `logabsdet` before, which would throw an error. I think I agree with you though. The reason why I added it is because of performance and numerical errors might lead to a singular Jacobian. I was thinking it could be useful instead return a very small number, still allowing AD to go through whatever else was included in the computation, e.g. ELBO in ADVI. This way, the parameters could potentially make their way "out of" the numerically unstable domain and optimization could run as normal. I say performance because I was under the impression that it's more expensive to have a `catch` vs. a `if -Inf` outside of the call? Either way, the user has to handle it. Would it be cheaper to warn the user but still use the `-Inf`?

_But_ I don't like it:) It's hacky. One the other hand it made LDA work, but for different reasons. In LDA we need Dirichlet distribution, whose bijector is using clamping (see `_clamp` in the `invlink` function). This lead of course lead to singular Jacobian. Even so, I could start out with singular Jacobian, run it for a while, and eventually it would become non-singular because we'd reach a domain were clamping was not necessary.

Reply:
Not very sure what's the best thing to do TBH. Was throwing an error the original behaviour? Maybe we should stick to it as this is an interface PR anyway, and solve this issue later. And I'm not very sure if returning `-Inf` would still make the trianing valid as there are no gradient through this function anymore right? Does it mean the gradient is wrong then?

Reply:
> Was throwing an error the original behaviour? Maybe we should stick to it as this is an interface PR anyway, and solve this issue later. And I'm not very sure if returning -Inf would still make the trianing valid as there are no gradient through this function anymore right? 

Well, before we were using `logabsdet` which throws an error because it uses LU-decomposition to compute the determinant, which is only possible if `det(J) != 0`. Now we're using `log(abs(det(J)))` instead, which means that we're either going to have an error when calling `log(0)` or return `-Inf`; either I think I'm fine with.

Returning `-Inf` doesn't help with computing the gradient, but in whatever computation you could do
```julia
if logabsdetjac(b, x) == -Inf
    # deal with it as you please
    ...
end
```
compared to catching the error and dealing with the error instead.

Or we could go back to using `logabsdet`.

> Does it mean the gradient is wrong then?

Computing the jacobian for the Dirichlet distribution potentially involves clamping, which makes jacobian singular and thus `logdet = -Inf`. So the gradient of this is of course not well-defined, if that's what you mean?